### PR TITLE
tools: scripts: Makefile: Added altera build

### PR DESCRIPTION
**Pull request by scuciurean**
The makefile now has the possibility to build and upload the no-Os software, supporting Intel and Xilinx carriers.
Added environment scripts that are used when building with an SDK and updated the project's source makefile.
Added bsp handling so it will be generated only once (issue #350).

### CODE REVIEW

File: tools/scripts/Makefile [hunk 1]
```diff
@@ -1,7 +1,14 @@
-export PROCESSOR_CORE 
+#------------------------------------------------------------------------------
```
Comment (mirceacaprioru) [inline tools/scripts/Makefile]: this modification is not related to altera build, i would do it in a separate commit

File: tools/scripts/Makefile [hunk 2]
```diff
@@ -29,25 +42,38 @@ MUTE = @
...
+#                            COMMON COMPILER FLAGS                             
+#------------------------------------------------------------------------------
 CFLAGS = -Wall 								\
 	 -Wmissing-field-initializers					\
 	 -Wclobbered 							\
 	 -Wempty-body 							\
 	 -Wignored-qualifiers 						\
 	 -Wmissing-parameter-type					\
+	 -Wno-format  							\
```
Comment (mirceacaprioru) [inline tools/scripts/Makefile]: not related to this commit

**Comment by buha** [comment]
I tested this on Xilinx only and these work:
- Upload functionality works for me
- 2nd `make` does indeed only build the source files

Some things which could perhaps be improved before the merge:
- Running `make` consecutively should only build modified source files in `build/` dir. It currently builds everything even if there were no changes.
- Running `make` after a `make clean` on a bare project without any `.hdf` file yields this output. The `cat ... arch.txt` line shouldn't be there.
```
$ make
[11:32:38] Evaluating hardware : 
[11:32:38] Platform not found
cat: /home/dberghe/Work/no-OS/projects/adrv9009/../../projects/adrv9009/build/tmp/arch.txt: No such file or directory
[11:32:38] Platform : none
[11:32:38] Can't generate the bsp
make: *** [../../tools/scripts/Makefile:274: bsp] Error 1
```

File: tools/scripts/Makefile [hunk 3]
```diff
@@ -372,10 +377,24 @@ post-build:
 		mv -f sw.map $(TEMP_DIR);				\
 	fi;
 
-xilinx:	xil-pre-cleanup eval-hardware pre-build xilinx-bsp
-	$(call copy_srcs,xilinx)
-	@$(MAKE) -s compile
-	rm -rf $(BUILD_DIR)/SDK.log
+# Upload the bitstream and elf file to the board
+.SILENT:run
+run: eval-hardware
```
Comment (buha) [inline tools/scripts/Makefile]: `run` should depend on the source compilation rule, otherwise it might upload an old binary

**Comment by scuciurean** [comment]
Changelog -> v2:

1. Splitted altera build commit.

File: tools/scripts/Makefile (hunk 3 above)
Comment (scuciurean) [reply tools/scripts/Makefile]: Doing this means that the user can run the binary only once per build.

File: tools/scripts/platform/intel/environment.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+:$1/$2/embedded/host_tools/altera/mkimage\
+:$1/$2/embedded/host_tools/altera/mkpimag\
+:$1/$2/embedded/host_tools/altera/device_tee\
+:$1/$2/embedded/host_tools/altera/diskutils\
+:$1/$2/embedded/host_tools/altera/imagecat\
+:$1/$2/embedded/host_tools/altera/securebot\
+:$1/$2/embedded/host_tools/gnu/dtc\
+:$1/$2/embedded/ds-5/sw/gcc/bin\
+:$1/$2/embedded/ds-5/sw/ARMCompler5.06u1/bin\
+:$1/$2/embedded/ds-5/bin"
```
Comment (dbogdan) [inline tools/scripts/platform/intel/environment.sh]: Let's include at this point only the mandatory tools. Initially, we will support only Nios2 platforms.

File: tools/scripts/platform/xilinx/environment.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+:$1/Vivado/$2/bin\
+:$1/SDK/$2/bin\
+:$1/SDK/$2/gnu/microblaze/lin/bin\
+:$1/SDK/$2/gnu/arm/lin/bin\
+:$1/SDK/$2/gnu/microblaze/linux_toolchain/lin64_le/bin\
+:$1/SDK/$2/gnu/aarch32/lin/gcc-arm-linux-gnueabi/bin\
+:$1/SDK/$2/gnu/aarch32/lin/gcc-arm-none-eabi/bin\
+:$1/SDK/$2/gnu/aarch64/lin/aarch64-linux/bin\
+:$1/SDK/$2/gnu/aarch64/lin/aarch64-none/bin\
+:$1/SDK/$2/gnu/armr5/lin/gcc-arm-none-eabi/bin\
```
Comment (dbogdan) [inline tools/scripts/platform/xilinx/environment.sh]: Same comment for the Altera tools is valid for the Xilinx ones.

File: tools/scripts/Makefile (hunk 2 above)
Comment (dbogdan) [reply tools/scripts/Makefile]: I still see unrelated changes...

Comment (scuciurean) [reply tools/scripts/Makefile]: This flag is used by default in the intel sdk

**Comment by scuciurean** [comment]
Changelog -> V2:

1. Stripped down the environment scripts.

**Comment by buha** [comment]
I did some research on how `rsync` might be used instead of `cp` to copy only modified files (newer filesystem timestamp) from repo to `build/` dir. It turns out a simple rule like this will do this:
```
rsync-srcs:                                                                      
    printf '%s\n' $(subst LOCAL_PLATFORM,$(PLATFORM),$(SRCS)) >  $(BUILD_DIR)/srclist.txt
    printf '%s\n' $(subst LOCAL_PLATFORM,$(PLATFORM),$(INCS)) >> $(BUILD_DIR)/srclist.txt
    rsync -t --no-relative --files-from=$(BUILD_DIR)/srclist.txt / $(BUILD_DIR)/app/src
```
rsync exists on all linuxes. As a first step, we could do a simple check to use `copy-srcs` on Windows and `rsync-srcs` on Linux. Then, if there is interest, I'm sure it can be implemented on Windows too.

I still think this feature is a must. The whole point of `make` is to build only modified sources and we need this in our project.

**Comment by scuciurean** [comment]
Changelog -> V4:

1. Split in more commits
2. Fixed conflicts

**Comment by scuciurean** [comment]
Changelog -> V5:

1. Renamed platform folder from intel to altera.

**Comment by dbogdan** [comment]
If you decided to do the rename, please update tools/scripts/platform/intel/environment.sh too.

**Comment by dbogdan** [comment]
Otherwise, it looks good.

**Comment by scuciurean** [comment]
> If you decided to do the rename, please update tools/scripts/platform/intel/environment.sh too.

I forgot to push the changes
